lib/iris/experimental/regrid.py

```diff
@@ -827,43 +874,59 @@ def regrid_weighted_curvilinear_to_rectilinear(src_cube, weights, grid_cube):
     # Get the target grid cube x and y dimension coordinates.
     tx, ty = get_xy_dim_coords(grid_cube)
 
-    if sx.units.modulus is None or sy.units.modulus is None or \
-            sx.units != sy.units:
+    #
+    # FOR NOW: minimal hacks to allow non-latlon grids
+    #
+    if sx.units != sy.units:
         msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
-            'have units of degrees or radians.'
+            'have the same units.'
         raise ValueError(msg.format(sx.name(), sy.name()))
 
-    if tx.units.modulus is None or ty.units.modulus is None or \
-            tx.units != ty.units:
+    if tx.units != ty.units:
         msg = 'The target grid cube x ({!r}) and y ({!r}) coordinates must ' \
-            'have units of degrees or radians.'
+            'have the same units.'
         raise ValueError(msg.format(tx.name(), ty.name()))
 
-    if sx.units != tx.units:
-        msg = 'The source cube and target grid cube must have x and y ' \
-            'coordinates with the same units.'
-        raise ValueError(msg)
+#    if sx.units != tx.units:
+#        msg = 'The source cube and target grid cube must have x and y ' \
+#            'coordinates with the same units.'
+#        raise ValueError(msg)
 
     if sx.ndim != sy.ndim:
         msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
             'have the same dimensionality.'
         raise ValueError(msg.format(sx.name(), sy.name()))
 
-    if sx.ndim != 2:
-        msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
-            'be 2D auxiliary coordinates.'
-        raise ValueError(msg.format(sx.name(), sy.name()))
+#    if sx.ndim != 2:
+#        msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
+#            'be 2D auxiliary coordinates.'
+#        raise ValueError(msg.format(sx.name(), sy.name()))
 
     if sx.coord_system != sy.coord_system:
         msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
             'have the same coordinate system.'
         raise ValueError(msg.format(sx.name(), sy.name()))
 
-    if sx.coord_system != tx.coord_system and \
-            sx.coord_system is not None and \
-            tx.coord_system is not None:
-        msg = 'The source cube and target grid cube must have the same ' \
-            'coordinate system.'
+    if src_cube.coord_dims(sx) != src_cube.coord_dims(sy):
+        msg = 'The source cube x ({!r}) and y ({!r}) coordinates must ' \
+            'map onto the same cube dimensions.'
+        raise ValueError(msg.format(sx.name(), sy.name()))
+
+    if weights is None:
+        weights = np.ones(sx.shape)
+    if weights.shape != sx.shape:
```

Comment:
This requires that `sx` and `sy` have the same shape, which is fine in the multiple '1D' horizontal coords case. But what about when we have '2D' horizontal coords - can we now only regrid square cubes?